impulse/lfsr.nim

```diff
@@ -0,0 +1,231 @@
+## used to generate pseudo-random boolean sequences. It supports both Fibonacci
+## and Galois LFSRs.
+##
+## LFSRs used in many digital communication systems (including, for example LTE
+## and 5GNR). For more information see:
+## https://simple.wikipedia.org/wiki/Linear-feedback_shift_register
+##
+## Notes:
+## - This code is heavily based on Nikesh Bajaj's pylfsr, which can be
+## found in https://pylfsr.github.io
+## - This implementation is not optimized for performance, but for simplicity.
+##   It would be relatively trivial to implement a much faster version by
+##   operating on the bits directly, rather than using tensors.
+
+import arraymancer
+import std/strformat
+
+type LFSR_TYPE* = enum
+  ## LFSR types (see https://simple.wikipedia.org/wiki/Linear-feedback_shift_register)
+  fibonacci, galois
+
+type LFSR_INIT_STATE_TYPE* = enum
+  ## Common LFSR Initial States
+  ## While it is possible to set any LFSR initial `state` by passing a
+  ## Tensor[bool] to `initLFSR`, for convenience it is also possible
+  ## to pass one of these enum values as the initial `state`:
+  ## - `single_true`: all bits set to `false` except the LSB (i.e. the last one)
+  ## - `all_true`: all bits set to `true`
+  single_true, all_true
+
+type LFSR* = object
+  ## LFSR object used to generate fibonacci or galois pseudo random sequences
+  ## To use it first create the object using `initLFSR` and then call either
+  ## the `next` procedure to get the sequence values one by one or `generate`
+  ## to generate multiple values in one go.
+  taps*: Tensor[int]
+  conf*: LFSR_TYPE
+  state*: Tensor[bool]
+  init_state: Tensor[bool]
+  verbose*: bool
+  counter_starts_at_zero*: bool
+  outbit*: bool
+  count*: int
+  seq_bit_index: int
+  sequence: Tensor[bool]
+  feedbackbit: bool
+
+proc initLFSR*(taps: typeof(LFSR.taps),
+    conf = fibonacci,
+    state: Tensor[bool],
+    verbose = false,
+    counter_starts_at_zero = true): LFSR =
+  ## Initialize LFSR with given feedback polynomial and initial state
+  ##
+  ## Inputs:
+  ## - taps: Feedback polynomial taps as a Tensor[int] of exponents in
+  ##         descending order. Note that exponent 0 can be omitted, since it is
+  ##          always implicitly added. For example, to use the `x^5 + x^3 + 1`
+  ##         polinomial you can set taps to `[5, 3].toTensor` or to
+  ##         `[5, 3, 0].toTensor`.
+  ## - state: Initial state of the LFSR as a Tensor[bool] of size equal to the
+  ##          highest exponent in taps (which must be its first element)
+  ## - conf: LFSR type as an enum (`fibonacci` or `galois`)
+  ## - verbose: Enable it to print additonal logs
+  ## - counter_starts_at_zero: Start the count from 0 or 1 (defaults to `true`)
+  ##
+  ## Return:
+  ## - Ready to use LFSR object
+
+  # Remove the last value from taps if it is a zero
+  let taps = if taps[taps.size - 1] == 0: taps[_..^2] else: taps
+  if taps.size > 1 and taps[0] < taps[taps.size - 1]:
```

Comment:
Doesn't this mean for multiple polynomials an ordering like `[7, 3, 5, 2]` would pass the check? If so one could use `isSorted` (https://nim-lang.github.io/Nim/algorithm.html#isSorted%2CopenArray%5BT%5D)